src/attributes/codegen.md

```diff
@@ -133,7 +134,8 @@ Feature     | Implicitly Enables | Description
 
 #### `wasm32` or `wasm64`
 
-This platform allows `#[target_feature]` to be applied to both safe and
+Executing code with unsupported features is allowed (i.e., is not UB) on this platform.
+Hence this platform allows `#[target_feature]` to be applied to both safe and
```

Comment:
This may want to be reworded slightly, since I think we still want to reserve the right to say that if you actually execute undefined instructions that's UB. That being said the target inherently provides no mechanism or means by which to do this.

Put another way "executing code with unsupported feautres" is not necesarily explicitly safe on wasm platforms, it's just not inherently possible in the platforms itself.

Basically I want to leave us room to handle [feature detection](https://github.com/WebAssembly/feature-detection/blob/main/proposals/feature-detection/Overview.md) at some point in the future. Even then though that will be implemented in a way that you won't be able to get it wrong so it won't be `unsafe` (just sort of inherently with the wasm platform).

I dunno this is definitely sort of lawyer-y, but I think ideally this would be reworded a bit to say that `#[target_feature]` can be applied to safe functions because it's not possible to execute unknown instructions on wasm.

Reply:
> This may want to be reworded slightly, since I think we still want to reserve the right to say that if you actually execute undefined instructions that's UB. That being said the target inherently provides no mechanism or means by which to do this.

UB is a Rust concept here, not a target concept. So this is about explaining what you are or are not allowed to do *in Rust* when compiling to this target.

In Rust it is definitely possible to write code that ends up executing unknown instructions, even when compiling to wasm. My understanding is that this is guaranteed to trap, but that is not the same as "it's impossible to do so".

Reply:
> My understanding is that this is guaranteed to trap, but that is not the same as "it's impossible to do so".

AFAIK it causes the wasm module to fail to load at all if the simd instructions are not optimized away. The wasm runtime needs to know about all used instructions in a wasm module to be able to parse and typecheck them.

Reply:
Ok, I'll leave  this to you then

Reply:
To be clear, I am open to suggestions. :)

Reply:
@alexcrichton this paragraph got reworded, what do you think about the latest version?

> `#[target_feature]` may be used with both safe and
[`unsafe` functions][unsafe function] on Wasm platforms. It is impossible to
cause undefined behavior via the `#[target_feature]` attribute because
attempting to use instructions unsupported by the Wasm engine will fail at load
time without the risk of being interpreted in a way different from what the
compiler expected.

Reply:
Sure!